docs/pallas/tpu/distributed.md

```diff
@@ -0,0 +1,1453 @@
+
+    @pl.when(phase == RIGHT)
+    def _():
+      pltpu.semaphore_wait(left_capacity_sem, 1)
+
+
+out_shape = (
+    jax.ShapeDtypeStruct(
+        (outer_block_size[0], outer_block_size[1]), jnp.float32
+    ),
+    # Shape: [working/recv, block[0], block[1]]
+    jax.ShapeDtypeStruct(
+        (2, outer_block_size[0], outer_block_size[1]), jnp.float32
+    ),  # hbm_scratch
+)
+
+grid_spec = pltpu.PrefetchScalarGridSpec(
+    num_scalar_prefetch=0,
+    in_specs=[
+        pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+    ],
+    out_specs=[
+        pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+        pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+    ],
+    grid=(num_devices, 2),
+    scratch_shapes=(
+        [pltpu.SemaphoreType.DMA] * 5
+        + [pltpu.SemaphoreType.REGULAR] * 2  # Capacity semaphores
+    ),
+)
+
+
+def pallas_reduce_scatter(input_arr):
+  input_arr = input_arr.reshape(
+      num_devices, outer_block_size[0], outer_block_size[1]
+  )
+  return pl.pallas_call(
+      reduce_scatter_kernel,
+      out_shape=out_shape,
+      grid_spec=grid_spec,
+      compiler_params=dict(mosaic=dict(collective_id=0)),
+  )(input_arr)[0]
+
+
+pallas_result = jax.jit(
+    shard_map.shard_map(
+        pallas_reduce_scatter,
+        mesh=mesh,
+        in_specs=P(None, 'x'),
+        out_specs=P('x', None),
+        check_rep=False,
+    )
+)(input_arr)
+
+pallas_result = jax.block_until_ready(pallas_result)
+```
+
+```{code-cell} ipython3
+---
+executionInfo:
+  elapsed: 768
+  status: ok
+  timestamp: 1722904808851
+  user:
+    displayName: Justin Fu
+    userId: '17543197034567316452'
+  user_tz: 420
+id: cTEyiMDyx9Y0
+outputId: 1de26695-3713-430e-9ab4-4ea646691680
+---
+# Now we compare our result to XLA.
+def lax_reduce_sum_scatter(x):
+  x = x.reshape(num_devices, outer_block_size[0], outer_block_size[1])
+  return lax.psum_scatter(x, 'x')
+
+
+xla_result = jax.jit(
+    shard_map.shard_map(
+        lax_reduce_sum_scatter,
+        mesh=mesh,
+        in_specs=P(None, 'x'),
+        out_specs=P('x', None),
+    )
+)(input_arr)
+
+print('Input:', input_arr.shape, input_arr[::4, 0])
+print('Pallas Result:', pallas_result.shape, pallas_result[::4, 0])
+print('lax.psum_scatter Result:', xla_result.shape, xla_result[::4, 0])
+print(
+    'Difference |Pallas - lax.psum_scatter|:',
+    jnp.max(jnp.abs(pallas_result - xla_result)),
+)
+```
+
++++ {"id": "zz5AFbriliyv"}
+
+## Conclusion
+
+In this tutorial, we have covered a wide range of APIs and techniques for writing efficient distributed kernels on TPUs. This includes:
+
+- TPU topologies and ring communication patterns.
+- The remote DMA API (`make_async_remote_copy`).
+- Synchronization with DMA, Regular, and Barrier semaphores.
+- Double-buffering.
+- Bi-directional communication for bandwidth optimization.
+- Nested pipelining with `emit_pipeline`.
+- Implementations of `lax.ppermute` (right), `lax.all_gather`, `lax.psum`, `lax.psum_scatter`.
```

Comment:
I wonder if it's worth calling out that these are reference implementations that XLA can't async-decompose and schedule to get overlap, like it can with the HLO versions.

Reply:
Discussed offline. This seems like it's supported conditionally (for some collectives and on newer hardware), so I will leave this point out as there's a lot of edge cases.